### ISSUE TITLE
[red-knot] Context managers should use `try_call_dunder` infrastructure

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ruff#16371]
### Summary

`__enter__` and `__exit__` should be treated like all other dunder methods. However, while working on #16368, I noticed that we look them up manually (and potentially incorrectly) here …

https://github.com/astral-sh/ruff/blob/1be0dc6885317340acdd3427aa43acc05e7a43d3/crates/red_knot_python_semantic/src/types/infer.rs#L1621-L1622

… and call them manually (and potentially incorrectly) here:

https://github.com/astral-sh/ruff/blob/1be0dc6885317340acdd3427aa43acc05e7a43d3/crates/red_knot_python_semantic/src/types/infer.rs#L1660-L1661

This only leads to incorrect behavior in very exotic situations like the following, where we incorrectly raise a *"Object of type `Manager` cannot be used with `with` because it does not correctly implement `__enter__`"* diagnostic:

```py
from __future__ import annotations

class Target: ...

class SomeCallable:
    def __call__(self) -> Target:
        return Target()

class Descriptor:
    def __get__(self, instance, owner) -> SomeCallable:
        return SomeCallable()

class Manager:
    __enter__: Descriptor = Descriptor()

    def __exit__(self, exc_type, exc_value, traceback): ...

with Manager() as f:
    reveal_type(f)  # revealed: Target
```

Fixing this would involve using `try_call_dunder` after #16368 is merged, or using `static_member` + a manual descriptor call, if that is not possible.

**Comment by sharkdp** [comment]
The same is probably true for `__class_getitem__`:

https://github.com/astral-sh/ruff/blob/1be0dc6885317340acdd3427aa43acc05e7a43d3/crates/red_knot_python_semantic/src/types/infer.rs#L5220-L5221

https://github.com/astral-sh/ruff/blob/1be0dc6885317340acdd3427aa43acc05e7a43d3/crates/red_knot_python_semantic/src/types/infer.rs#L5238

**Comment by sharkdp** [comment]
And augmented assignments:

https://github.com/astral-sh/ruff/blob/1be0dc6885317340acdd3427aa43acc05e7a43d3/crates/red_knot_python_semantic/src/types/infer.rs#L2259-L2267

**Comment by sharkdp** [comment]
And `in`/`not in` tests:

**Comment by sharkdp** [comment]
… and subscript expressions (this is the last one, I promise):

https://github.com/astral-sh/ruff/blob/1be0dc6885317340acdd3427aa43acc05e7a43d3/crates/red_knot_python_semantic/src/types/infer.rs#L5220-L5238
